### DISCUSSION TITLE
4.4.16_0

### DISCUSSION THREAD

**Discussion by hymkor** [nyaosorg/nyagos#443] [General]
# Changes on 4.4.16_0

## Removed features and Deprecations

* Remove `nyagos.d/catalog/neco.lua`
* Remove Lua function: `nyagos.msgbox`
* Remove the feature to define `grep` as an alias of `findstr.exe` when `grep.exe` does not exist. It was the undocumented feature.

### Readline

![image](https://github.com/user-attachments/assets/f3fcb0b1-a370-4e26-8400-f3854281f654)

* PowerShell 7-style input prediction
    - When one or more characters are typed, the most recent entry in the history that starts with the typed characters is displayed inline (in blue italics).
    - Pressing `→` or `Ctrl-F` accepts the current prediction
    - Enabled by default. You can disable it with the launch option `--no-predict` or by setting `nyagos.option.prediction=false`
* Ctrl-P/N: save the modified entry when switching history, and restore when switching again, until Enter is pressed

### Lua extensions

* Implement `nyagos.getkeys()` which returns the input key as a [sequence][keyseq] like `"\027[A"` instead of `nyagos.getkey()` which returns unicode.
* Enhance functions available in `nyagos.key[KEY] = function(this)...end`
    * [key-sequences][keyseq] like `"\007"` and `"\027[A"` are avaliable as KEY.
    * Implement `this:eval(KEYSEQUNCE)` that calls the function assigned to [KEYSEQUNCE][keyseq].
    * Implement `this:repaint()` that updates the screen
    * `this.pos` and `this.text` are automatically updated when methods are invoked.
* Add the completion for `make`: Enable by `require "makefile-complete"`
* Add subcommand name completion for the jj command: Enable by requiring "complete-jj".
* Add citation feature for the gmnlisp command: Enable by `require "gmnlisp"`
    * Replaces `@(Lisp command)` with the result processed by gmnlisp.
    * If the command starts with `(`, it is executed using gmnlisp.exe.
* Add UNIX-style single quotation feature: Enable by requiring "sq2dq".
    * Replaces '..".."..' with "..\"..\"..".
    * Leaves "..'..'.." unchanged (single quotes within double quotes are not converted).

[keyseq]: https://learn.microsoft.com/en-us/windows/console/console-virtual-terminal-sequences#input-sequences

## Documents

* Fix the broken links caused by renaming readme.md to README.md ( Thx @HAYASHI-Masayuki )

# 4.4.16_0 での変更点

## 廃止・非推奨

* `nyagos.d/catalog/neco.lua` を削除
* Lua関数: `nyagos.msgbox` を削除
* grep.exe が存在しない時、勝手に findstr.exe を grep の別名にする動作を削除

### 一行入力

![image](https://github.com/user-attachments/assets/f3fcb0b1-a370-4e26-8400-f3854281f654)

* PowerShell 7 風の入力予測機能の実装
    - 一文字以上入力すると、その内容で始まる履歴の最新エントリをインラインで表示(青の斜体)
    - `→` もしくは `Ctrl-F` で今の予測表示内容を採用する
    - デフォルトでオン。起動オプション `--no-predict` もしくは `nyagos.option.prediction=false` でオフ
* Ctrl-P/N: 履歴を切り替えるときに変更したエントリを保存し、(Enterが入力されるまでは)再度切り替えたときに復元するようにした

### Lua 拡張

* キー入力の最初のコードの Unicode しか返さなくなっていた nyagos.getkey のかわりに、入力キーを`"\027[A"` といった[キーシーケンス]で返す nyagos.getkeys() を実装した
* `nyagos.key[KEY]=function(this)...end` の中で使える機能を拡充
    * KEY として従来の`"BACKSPACE"`, `"UP"` などの名前の他、`"\007"`, `"\027[A"` などの[キーシーケンス]も使えるようにした
    * `this:eval("キーシーケンス")` で[キーシーケンス]に設定された機能を呼び出せるようにした
    * 更新内容を画面に反映するメソッド`this:repaint()`を追加
    * 更新系のメソッドを呼び出した際に`this.pos`と`this.text` を自動的に更新するようにした
* `make` コマンドのエントリ補完を追加: `require "makefile-complete"` にて有効化
* `jj` コマンドのサブコマンド名補完を追加: `require "complete-jj"` にて有効化
* `gmnlisp` コマンドの引用機能を追加: `require "gmnlisp"` にて有効化
    - `@(Lispコマンド)` を gmnlisp で処理した結果に置換
    - コマンド先頭が `(` で始まっていた場合、gmnlisp.exe で実行
* UNIX風シングルクォーテーション機能を追加: `require "sq2dq"` にて有効化
    - `'..".."..'` を `"..\"..\".."` へ置換
    - `"..'..'.."` はそのまま (二重引用符内の一重引用符は変換しない)

[キーシーケンス]: https://learn.microsoft.com/en-us/windows/console/console-virtual-terminal-sequences#input-sequences

## ドキュメント

* readme.md → README.md でのファイル名変更でドキュメント間のリンクが切れてしまっていた点の修正 ( Thx @HAYASHI-Masayuki )

<hr /><em>This discussion was created from the release <a href='https://github.com/nyaosorg/nyagos/releases/tag/4.4.16_0'>4.4.16_0 (pre-release)</a>.</em>

**Comment by hymkor** [comment]
This version was pre-released on Oct.~~24~~ <ins>13</ins>, 2024.
If no problems are not found in one or two weeks, then pre-released flag will be reset and scoop-installer is avaliable for the new version.

本バージョンは 2024-10-~~24~~ <ins>13</ins> にプレリリースしました。
1~2週間様子を見て、特に問題なければ、プレリリースフラグをオフにして、scoop でも更新できるようにします。

**Reply by dai** [reply]
Oct.14, 2024?

細かくてすいませんｗ

**Reply by hymkor** [reply]
ありがとうございます

**Comment by hymkor** [comment]
The time the pre-release flag will be cleared comes soon, and does anyone who considers the PowerShell 7-style input prediction should be disabled on default, exist ?

ぼちぼち、プレリリースフラグを解除する時期が来てますが、PowerShell 7 風の入力予測機能はデフォルトでオフの方がよいと考える人とかいらっしゃいますでしょうか？

**Reply by dai** [reply]
受信ボックスに🔵が付いてたので開いたけど、酔っ払ってしまっているので後で良く見ます。
7風の入力予測がなんだっけ？レベルですｗ
ごめんなさいごめんなさい🙏

**Reply by hymkor** [reply]
これです

![](https://private-user-images.githubusercontent.com/3752189/376028906-f3fcb0b1-a370-4e26-8400-f3854281f654.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjk5NDcwNjAsIm5iZiI6MTcyOTk0Njc2MCwicGF0aCI6Ii8zNzUyMTg5LzM3NjAyODkwNi1mM2ZjYjBiMS1hMzcwLTRlMjYtODQwMC1mMzg1NDI4MWY2NTQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTAyNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDEwMjZUMTI0NjAwWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9Y2NkYTI1YzlkNmYwMTQ2NTRmZjExZjg0MDc3MGVkZGRjZTU5NmYxYjVlYTc1MTdkZGQ5YTFkNjgzOWRiYjBhOSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.spkJITtYeNEQ4hNK-w4tDL1e48URW56B_gRxUTHx34I)

今の入力中のテキストで始まる直近の履歴を自動で青字で表示する機能です。
（採用したいときは `→` もしくは `Ctrl-F` を押下）

便利ではあるんですが、画面がやはゴチャゴチャします。Ctrl-R とか Ctrl-P など従来の履歴参照を使いこなしている人にとっては、補完か履歴参照か一瞬混乱したりするので、デフォルトの設定でオフにする選択肢も検討していました。

**Reply by dai** [reply]
リリースありがとうございます。

デフォルトオンのままユーザーがopt-outするほうが、つまりリリース版（4.4.16_0）の通りで良いと思います。

個人的にはNYAGOSでの履歴は `Ctrl + P` をよく使ってますが、この機能では青文字斜体で表示されるためゴチャゴチャせず便利に使えてます。

// 頭で再現しても実際に手を動かさないとわからなかったため、手で試してからコメントをさせていただきました。

**Comment by hymkor** [comment]
The pre-released flag was cleared. You can upgrade to 4.4.16_0 with `scoop update nyagos` now.

pre-released フラグをクリアしました。`scoop update nyagos` で 4.4.16_0 に更新できるようになっています。

**Comment by hymkor** [comment]
4.4.17\_0 をリリースしましたので、本 discussion をクローズします。
